README.md
# k-rte

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/k-rte?sort=semver&label=latest)](https://github.com/kalisio/k-rte/releases)
[![Build Status](https://app.travis-ci.com/kalisio/k-rte.svg?branch=master)](https://app.travis-ci.com/kalisio/k-rte)

A [Krawler](https://kalisio.github.io/krawler/) based service to download generation data from the French gelectrical network operated by RTE.

## Description

The **k-rte** job allow to scrape generation data from the [RTE API](https://data.rte-france.com/). The downloaded data are stored within a [MongoDB](https://www.mongodb.com/) database and more precisely in 2 collections:
* the `rte-generation` collection stores the production units generation data
* the `rte-units` collection stores the production units positions

All records are stored in [GeoJson](https://fr.wikipedia.org/wiki/GeoJSON) format.

The project consists in 2 jobs:
* the `units` job scrapes the available production units according a specific cron expression. By default, every day at midnight.
* the `generation` job scrapes the generation data according a specific cron expression. By default every hour.

As the RTE API does not provide the location of the production units, the default `units` job simply read data from input files providing plant location and associated production units (taken from https://github.com/ewoken/nuclear-monitor), assuming all production units of the plant have the same location. You can update the file whenever production units are updated. If you have a better source of production units locations you should customise the job to use it.

## Configuration

### Production units

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/rte` |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

### Generation data

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/rte` |
| `TTL` | The observations data time to live. It must be expressed in seconds and the default value is `604 800` (7 days) | 
| `CLIENT_ID` | Client ID of your application in the RTE API, required to generate the authentication token. |
| `CLIENT_SECRET` | Client secret of your application in the RTE API, required to generate the authentication token. |
| `PRODUCTION_TYPE_FILTER` | Comma-separated list of production type to be read (please refer to RTE API documentation for details). |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

## Deployment

We personally use [Kargo](https://kalisio.github.io/kargo/) to deploy the service.

## Contributing

Please refer to [contribution section](./CONTRIBUTING.md) for more details.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE) for details

CHANGELOG.md
Please refer to GitHub issues for changelog: https://github.com/kalisio/k-rte/issues.


jobfile-units.js
import _ from 'lodash'
import moment from 'moment'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/rte'

// FIXME: now we only have information about nuclear plant/reactors
export default {
  id: 'rte-units',
  store: 'memory',
  options: {
    workersLimit: 1,
    faultTolerant: true,
  },
  tasks: [{
    id: 'reactors.csv',
    type: 'store',
    options: {
      store: 'fs'
    }
  }],
  hooks: {
    tasks: {
      after: {
        readCSV: {
          headers: true
        },
        apply: {
          function: (item) => {
            const plants = _.get(item, 'plants', [])
            delete item.plants
            const units = _.get(item, 'data', [])
            _.forEach(units, (unit) => {
              // Find owing plant to get location and other useful information
              const plant = _.find(plants, { id: unit.plantId })
              if (plant) {
                _.merge(unit, _.omit(plant, ['id', 'name']))
              }
              // Convert some properties
              _.forOwn(unit, (value, key) => {
                if (key.endsWith('_MW')) unit[key] = _.toNumber(unit[key])
                if (key.endsWith('Date')) unit[key] = moment.utc(unit[key]).toDate()
              })
            })
            console.log('Found ' + units.length + ' production units')
          }
        },
        convertToGeoJson: {
          latitude: 'lat',
          longitude: 'long'
        },
        updateMongoCollection: {
          collection: 'rte-units',
          filter: { 'properties.eicCode': '<%= properties.eicCode %>' },
          upsert: true,
          chunkSize: 256
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: [{
          id: 'memory'
        }, {
          id: 'fs', options: { path: __dirname }
        }],
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createMongoCollection: {
          clientPath: 'taskTemplate.client',
          collection: 'rte-units',
          indices: [
            [{ 'properties.eicCode': 1 }, { unique: true }], 
            { geometry: '2dsphere' }
          ]
        },
        readCSV: {
          key: 'plants.csv',
          store: 'fs',
          headers: true,
          dataPath: 'data.taskTemplate.plants'
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      }
    }
  }
}


jobfile-generation.js
import _ from 'lodash'
import moment from 'moment'

const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/rte'
const ttl = +process.env.TTL || (7 * 24 * 60 * 60)  // duration in seconds
const typeFilter = process.env.PRODUCTION_TYPE_FILTER ? process.env.PRODUCTION_TYPE_FILTER.split(',') : ''

export default {
  id: 'rte-generation',
  store: 'memory',
  options: {
    workersLimit: 1
  },
  tasks: [{
    id: 'generation',
    type: 'http',
    options: {
      url: 'https://digital.iservices.rte-france.com/open_api/actual_generation/v1/actual_generations_per_unit',
      oauth: {
        url: 'https://digital.iservices.rte-france.com/token/oauth',
        client_id: process.env.CLIENT_ID,
        client_secret: process.env.CLIENT_SECRET,
        method: 'client_secret_basic'
      },
      // It seems that we cannot request less than two days as
      // the API only seems to take date and not time into account
      // so request from yesterday to tomorrow
      start_date: moment.utc().subtract(1, 'day').startOf('day').format(),
      end_date: moment.utc().add(1, 'day').startOf('day').format()
    }
  }],
  hooks: {
    tasks: {
      before: {
        OAuth: {},
        createMongoAggregation: {
          dataPath: 'data.mostRecentData',
          collection: 'rte-generation',
          pipeline: [
            { $sort: { 'properties.eicCode': 1, time: 1 } },
            {
              $group:
                {
                  _id: "$properties.eicCode",
                  time: { $last: "$time" }
                }
            }
          ],
          allowDiskUse: true
        }
      },
      after: {
        readJson: {},
        apply: {
          function: (item) => {
            const units = _.get(item, 'units', [])
            console.log('Seeking generation data for ' + units.length + ' units')
            delete item.units
            const mostRecentData = _.get(item, 'mostRecentData', [])
            let generation = _.get(item, 'data.actual_generations_per_unit', [])
            // Filter required production types
            if (typeFilter) {
              generation = generation.filter(data => typeFilter.includes(_.get(data, 'unit.production_type')))
            }
            let features = []
            _.forEach(generation, (data) => {
              const eicCode = _.get(data, 'unit.eic_code')
              // Match unit using name for now
              const unit = _.find(units, unit => _.get(unit, 'properties.eicCode') === eicCode)
              if (unit) {
                const feature = {
                  type: 'Feature',
                  properties: _.pick(unit.properties, ['eicCode', 'name']),
                  geometry: unit.geometry
                }
                // Now keep track of newer values
                const latestData = _.find(mostRecentData, latestData => latestData._id === eicCode)
                const values = _.get(data, 'values', [])
                values.forEach(record => {
                  const time = moment.utc(record.end_date)
                  // Check if newer
                  if (latestData && time.isSameOrBefore(moment.utc(latestData.time))) return
                  // If so push it
                  features.push(_.merge({
                    time: time.toDate(),
                    properties: {
                      power: _.toNumber(record.value)
                    }
                  }, feature))
                })
              }
            })
            if (features.length > 0) console.log('Found ' + features.length + ' new generation data')
            else console.log('No new generation data found')
            item.data = features
          }
        },
        writeMongoCollection: {
          collection: 'rte-generation',/*
          transform: {
            mapping: { 'properties.measureDate': 'time' },
            omit: [ 'properties.location' ],
            unitMapping: { time: { asDate: 'utc' } } 
          },*/
          chunkSize: 256
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: { id: 'memory' },
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        readMongoCollection: {
          clientPath: 'taskTemplate.client',
          collection: 'rte-units',
          dataPath: 'data.taskTemplate.units'
        },
        createGenerationCollection: {
          hook: 'createMongoCollection',
          clientPath: 'taskTemplate.client',
          collection: 'rte-generation',
          indices: [
            [{ time: 1, 'properties.eicCode': 1 }, { unique: true }],
            { 'properties.power': 1 },
            { 'properties.eicCode': 1, 'properties.power': 1, time: -1 },
            [{ time: 1 }, { expireAfterSeconds: ttl }], // days in s
            { geometry: '2dsphere' }                                                                                                              
          ],
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory']
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory']
      }
    }
  }
}
